File: src/routes/velocity-cache.ts
/**
 * In-memory cache for `/api/velocity` results.
 *
 * `computeVelocity` does a Nextcloud fetch + Kalman + segmentation +
 * OSM enrichment + biometric joins; on a data-rich day it takes
 * 5–10 seconds. Most of that work is deterministic for a given
 * `(user_id, date, tz)`, and the user typically revisits the same
 * day several times during a session (tab switches, chevron
 * navigation, refreshes). Caching the result in-process turns
 * those repeat views into ~50ms reads.
 *
 * Design choices:
 *
 *   - **Per-pod only.** Cleared on every process restart, which
 *     means a deploy automatically invalidates everything — no
 *     schema-version tag, no stale-cache-after-logic-change risk.
 *     The trade-off: cold cache after each deploy, so the first
 *     view of any day after deploy still pays the full compute.
 *     Nothing changes the pipeline's answer WITHOUT a restart any
 *     more: the verified-core master toggle was the one exception
 *     and it was removed with the TS arm (#975). Anything that
 *     reintroduces live switching must clear this cache too.
 *
 *   - **Short TTL (5 min).** Today's date keeps accumulating new
 *     Owntracks pushes, and Fitbit sleep sync can land any time.
 *     A 5-minute window is short enough that the user sees fresh
 *     data within a fix-or-two of it arriving, and long enough
 *     that a typical session of tab-switching benefits.
 *
 *   - **LRU eviction at 32 entries.** Single user covers maybe a
 *     month of recently-visited days; 32 leaves headroom without
 *     unbounded growth.
 *
 *   - **In-flight dedup.** Two concurrent requests for the same
 *     key share a single computation. Without this, opening the
 *     dashboard in two tabs would trigger two parallel
 *     OSM-enrichment runs hitting the same DB rows.
 *
 *   - **Logs hit/miss to stdout.** So we can confirm the cache is
 *     working from `kubectl logs` without instrumenting the
 *     frontend.
 */

import type { VelocityResult } from "../geo/velocity.js";

interface CacheEntry {
	result: VelocityResult;
	cachedAtMs: number;
}

const TTL_MS = 5 * 60 * 1000;

/** TTL for a day that is still happening.
 *
 *  A finished day's classification is settled: replaying it tomorrow gives the
 *  same answer, so five minutes of staleness costs nothing. A day *in progress*
 *  is a different object — every new fix can change it, and the pipeline's
 *  verdict on an unfinished journey is provisional by construction. One minute
 *  into a tube ride the rail passes have too little track to identify the line,
 *  so the leg reads as an unidentified vehicle; two minutes later it is a train
 *  on the Metropolitan Line. Cache the first answer for five minutes and the
 *  user stares at a superseded verdict long after the pipeline would have
 *  corrected itself (the 2026-07-12 report).
 *
 *  The cost is affordable precisely because the day is partial: today's compute
 *  runs in a few hundred milliseconds mid-morning, not the 5–10 s a full
 *  data-rich day takes — there is simply less of it. So a live day gets a short
 *  TTL, still long enough to absorb tab-switching and chevron navigation. */
export const LIVE_TTL_MS = 60 * 1000;

const MAX_ENTRIES = 32;

const cache = new Map<string, CacheEntry>();
const inFlight = new Map<string, Promise<VelocityResult>>();

/** Bumped by {@link invalidateVelocityCache}. A compute that started under an
 *  older generation must not write its result into the cleared cache — see
 *  there for why clearing the map alone is not enough. */
let generation = 0;

/** Is `date` the day currently in progress, as the viewer experiences it?
 *
 *  The boundary that matters is the viewer's local midnight, not UTC's — at
 *  23:30 in London the UTC date has already rolled over, and treating the day
 *  as finished would freeze the evening's timeline an hour early. `tz`
 *  undefined mirrors the API's own fallback: the date is read as UTC.
 *
 *  `nowMs` is injectable so this is testable without freezing the clock. */
export function isLiveDay(date: string, tz: string | undefined, nowMs: number = Date.now()): boolean {
	const today = new Intl.DateTimeFormat("en-CA", {
		timeZone: tz ?? "UTC",
		year: "numeric",
		month: "2-digit",
		day: "2-digit",
	}).format(new Date(nowMs));
	return date === today;
}

/** Get a velocity result from the cache, or compute and cache it. Generic so
 *  the route can cache the result with request-scoped extras attached (the
 *  watch-battery series) without widening `VelocityResult` itself; one key
 *  always stores what its own `compute` returned.
 *
 *  `ttlMs` lets the caller shorten the window for a day still in progress —
 *  see {@link LIVE_TTL_MS}. */
export async function getVelocityCached<T extends VelocityResult>(
	key: string,
	compute: () => Promise<T>,
	ttlMs: number = TTL_MS,
): Promise<T> {
	const entry = cache.get(key);
	if (entry && Date.now() - entry.cachedAtMs < ttlMs) {
		// LRU bump: delete + re-insert so this key is now most-recent.
		cache.delete(key);
		cache.set(key, entry);
		console.log(`velocity-cache HIT ${key} age=${Math.round((Date.now() - entry.cachedAtMs) / 1000)}s`);
		return entry.result as T;
	}

	// In-flight dedup: if another request for the same key is already
	// computing, await its promise instead of starting a parallel run.
	const pending = inFlight.get(key);
	if (pending) {
		console.log(`velocity-cache JOIN ${key}`);
		return pending as Promise<T>;
	}

	console.log(`velocity-cache MISS ${key}`);
	const startedIn = generation;
	const promise = compute()
		.then((result) => {
			// Computed across an invalidation: this result came out of the engine
			// that was live when it started, which is exactly what the invalidation
			// was discarding. Return it to the caller who asked before the switch,
			// but do not seat it in the cache — that would re-admit a pre-switch
			// answer with a full fresh TTL.
			if (generation !== startedIn) {
				console.log(`velocity-cache STALE ${key} — spanned an invalidation, not stored`);
				return result;
			}
			// LRU eviction: if at cap, drop the oldest entry. Map
			// preserves insertion order; the first key is the oldest.
			if (cache.size >= MAX_ENTRIES) {
				const oldest = cache.keys().next().value;
				if (oldest !== undefined) cache.delete(oldest);
			}
			cache.set(key, { result, cachedAtMs: Date.now() });
			return result;
		})
		.finally(() => {
			// Only retire OUR slot. If an invalidation dropped this promise from
			// the map and a later caller registered a replacement compute for the
			// same key under the new engine, a bare delete would evict that live
			// entry and the caller after it would start a third run.
			if (inFlight.get(key) === promise) inFlight.delete(key);
		});

	inFlight.set(key, promise);
	return promise;
}

/** Clear the cache, dropping in-flight computes as well as seated entries.
 *
 * ⚠ `invalidateVelocityCache` IS GONE (#975), and so is the reason it existed.
 * Its only caller was the verified-core master toggle, which swapped the Lean
 * core for TS inside a live process — the one thing that changed what
 * `computeVelocity` returns WITHOUT a restart. With the toggle removed, the
 * cache's per-pod assumption holds again on its own: only a deploy changes the
 * answer, and a deploy restarts the pod. The invalidator is not unused so much
 * as unnecessary.
 *
 * ⚠ Anything that later changes the pipeline's answer under a running pod must
 * bring it back. This is the note that says so.
 *
 * The generation bump is what drops in-flight work: a compute that began before
 * the clear still completes and still returns to whoever was already awaiting
 * it, but its result is not seated afterwards. Underscored because production
 * has no caller — the tests are the only ones, and that is the point. */
export function _resetVelocityCache(): void {
	generation++;
	cache.clear();
	inFlight.clear();
}

File: rust/backend/src/velocity_cache.rs
//! The per-pod cache in front of `/velocity` (#982).
//!
//! Computing a day is a Nextcloud fetch, a Kalman pass, segmentation, OSM
//! enrichment and biometric joins — 5–10 s on a data-rich day — and
//! deterministic for a given `(user, date, tz, walkMatch)`. The user revisits
//! the same day several times in a session, so repeat views read from here.
//!
//! # ⚠ Nothing here decides anything
//!
//! How long an entry may be reused, and how many to keep, are
//! `Verified.VelocityCache`. This is the map, the eviction order and the
//! single-flight — which request shares a computation with which is
//! concurrency, not policy. [`Policy`] is what the caller carries across from
//! Lean; it is never computed in this file.
//!
//! # Per-pod, cleared by restart
//!
//! No schema-version tag and no invalidation hook: a deploy restarts the pod
//! and that is the invalidation. The cost is a cold cache after each deploy.
//!
//! ⚠ **`invalidateVelocityCache` and its generation counter are NOT ported, and
//! that is a deletion rather than an omission.** In the TypeScript they existed
//! for one caller — the verified-core master toggle, which swapped the Lean core
//! for TS inside a live process. That toggle went with the TS arm (#975), and
//! with it the only thing that changed an answer without a restart. Anything
//! that later changes the pipeline's answer under a RUNNING pod has to bring
//! both back; this paragraph is the note that says so.

use std::collections::HashMap;
use std::sync::{Arc, Mutex};

use anyhow::Result;
use serde_json::Value;

/// What Lean decided for this request. Carried, never computed here.
#[derive(Debug, Clone, Copy)]
pub struct Policy {
    /// `Verified.VelocityCache.ttlMsFor` — the window this key may be reused in.
    pub ttl_ms: i64,
    /// `Verified.VelocityCache.MAX_ENTRIES` — the LRU bound.
    pub max_entries: usize,
}

struct Entry {
    key: String,
    result: Value,
    cached_at_ms: i64,
}

/// The cache. One per process; `AppState` holds it behind an `Arc`.
#[derive(Default)]
pub struct VelocityCache {
    /// ⚠ Insertion-ordered, oldest first, because the eviction rule is "drop the
    /// oldest" and a `HashMap` has no oldest. A `Vec` rather than an ordered-map
    /// crate: the bound is 32, so a linear scan is free and the order is visible
    /// in the type instead of being a property of a dependency.
    entries: Mutex<Vec<Entry>>,
    /// One lock per in-flight key. See [`VelocityCache::get_or_compute`].
    locks: Mutex<HashMap<String, Arc<tokio::sync::Mutex<()>>>>,
}

impl VelocityCache {
    pub fn new() -> Self {
        Self::default()
    }

    /// A fresh entry for `key`, bumped to most-recent. `None` on miss or stale.
    ///
    /// ⚠ Freshness is Lean's — including the rule that an entry stamped in the
    /// FUTURE is stale. Plain `now - cached_at < ttl` reads a far-future entry
    /// as eternally fresh, which is the one staleness failure that never expires
    /// by itself.
    fn take_fresh(&self, key: &str, now_ms: i64, ttl_ms: i64) -> Result<Option<Value>> {
        let mut entries = self.entries.lock().expect("velocity cache mutex");
        let Some(i) = entries.iter().position(|e| e.key == key) else {
            return Ok(None);
        };
        if !crate::lean::velocity_cache_fresh(entries[i].cached_at_ms, now_ms, ttl_ms)? {
            // Drop it rather than leave it to be re-tested on every request.
            entries.remove(i);
            return Ok(None);
        }
        // LRU bump: move to the end, so this key is now the most recent.
        let e = entries.remove(i);
        let result = e.result.clone();
        entries.push(e);
        Ok(Some(result))
    }

    fn seat(&self, key: &str, result: &Value, now_ms: i64, max_entries: usize) {
        let mut entries = self.entries.lock().expect("velocity cache mutex");
        if let Some(i) = entries.iter().position(|e| e.key == key) {
            entries.remove(i);
        }
        // Evict from the front — oldest first — until there is room.
        while entries.len() >= max_entries.max(1) {
            entries.remove(0);
        }
        entries.push(Entry {
            key: key.to_string(),
            result: result.clone(),
            cached_at_ms: now_ms,
        });
    }

    /// Read `key`, or compute it and seat the result.
    ///
    /// # Single-flight, and the one way it differs from the TypeScript
    ///
    /// Two concurrent requests for the same key must not both run the compute —
    /// opening the dashboard in two tabs would otherwise fire two parallel OSM
    /// enrichment runs over the same rows. The TypeScript shares the PROMISE, so
    /// the second caller receives whatever the first produced, failure included.
    ///
    /// This takes a per-key lock instead, and the waiter re-reads the cache
    /// after acquiring it. Same guarantee about parallelism; the difference is
    /// on FAILURE, where a waiter recomputes rather than inheriting the first
    /// caller's error. That is the better answer for a transient fault and a
    /// worse one for a persistent it — bounded by the number of waiters, which
    /// is the number of tabs.
    pub async fn get_or_compute<F, Fut>(
        &self,
        key: &str,
        now_ms: i64,
        policy: Policy,
        compute: F,
    ) -> Result<Value>
    where
        F: FnOnce() -> Fut,
        Fut: std::future::Future<Output = Result<Value>>,
    {
        if let Some(hit) = self.take_fresh(key, now_ms, policy.ttl_ms)? {
            tracing::debug!(key, "velocity-cache HIT");
            return Ok(hit);
        }

        let lock = {
            let mut locks = self.locks.lock().expect("velocity lock map mutex");
            Arc::clone(locks.entry(key.to_string()).or_default())
        };
        let guard = lock.lock().await;

        // ⚠ RE-READ under the lock. The holder we just queued behind may have
        // seated exactly what this request wanted, and computing again would be
        // the parallel run the lock exists to prevent, merely serialised.
        let out = match self.take_fresh(key, now_ms, policy.ttl_ms)? {
            Some(hit) => {
                tracing::debug!(key, "velocity-cache JOIN");
                Ok(hit)
            }
            None => {
                tracing::debug!(key, "velocity-cache MISS");
                match compute().await {
                    Ok(result) => {
                        self.seat(key, &result, now_ms, policy.max_entries);
                        Ok(result)
                    }
                    // ⚠ A failed compute seats NOTHING. Caching an error would
                    // serve it for the whole TTL to requests that might have
                    // succeeded.
                    Err(e) => Err(e),
                }
            }
        };

        drop(guard);
        // Retire the slot only when nobody else holds it: the map's reference
        // plus ours is two, and a waiter that cloned under this same lock would
        // make it three. Checked while holding the map lock, so a joiner cannot
        // arrive between the count and the removal.
        {
            let mut locks = self.locks.lock().expect("velocity lock map mutex");
            if Arc::strong_count(&lock) == 2 {
                locks.remove(key);
            }
        }
        out
    }

    /// How many entries are seated. Test-only reach into the shape.
    pub fn len(&self) -> usize {
        self.entries.lock().expect("velocity cache mutex").len()
    }

    pub fn is_empty(&self) -> bool {
        self.len() == 0
    }
}
